src/expr/impl/src/scalar/case.rs

```diff
@@ -106,6 +107,132 @@ impl Expression for CaseExpression {
     }
 }
 
+/// With large scale of simple form match arms in case-when expression,
+/// we could optimize the `CaseExpression` to `ConstantLookupExpression`,
+/// which could significantly facilitate the evaluation of case-when.
+#[derive(Debug)]
+struct ConstantLookupExpression {
+    return_type: DataType,
+    arms: HashMap<ScalarImpl, BoxedExpression>,
```

Comment:
Another choice is `BTreeMap` and it should theoretically perform better than `HashMap` if the cardinality is relatively small. 😄 Would you mind doing a benchmark for that?

Reply:
Performed the benchmark on the fairly straightforward test I created with the PR, average execution time is ~62ms, which is slower than `HashMap`.

This could probably due to there is no default `Ord` implementation for `ScalarImpl`, which needs to be wrapped in a `DefaultOrdered<T>` in order to be inserted into / retrieved from `BTreeMap`.

And for larger input, I assume that `HashMap` will also have a better performance than `BTreeMap`.

Reply:
And I'm thinking about cost model, with which we could probably have more adaptive choice of this kind of optimization during runtime. 🤔️

P.S. We probably don't need very accurate statistics, just an estimate.